transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java

```diff
@@ -0,0 +1,279 @@
+        iovecArrayPool = new IovecArrayPool();
+    }
+
+    @Override
+    protected Queue<Runnable> newTaskQueue(int maxPendingTasks) {
+        return newTaskQueue0(maxPendingTasks);
+    }
+
+    private static Queue<Runnable> newTaskQueue0(int maxPendingTasks) {
+        // This event loop never calls takeTask()
+        return maxPendingTasks == Integer.MAX_VALUE? PlatformDependent.<Runnable>newMpscQueue()
+                : PlatformDependent.<Runnable>newMpscQueue(maxPendingTasks);
+    }
+
+    public void add(AbstractIOUringChannel ch) {
+        logger.trace("Add Channel: {} ", ch.socket.intValue());
+        int fd = ch.socket.intValue();
+
+        channels.put(fd, ch);
+    }
+
+    public void remove(AbstractIOUringChannel ch) {
+        logger.trace("Remove Channel: {}", ch.socket.intValue());
+        int fd = ch.socket.intValue();
+
+        AbstractIOUringChannel old = channels.remove(fd);
+        if (old != null && old != ch) {
+            // The Channel mapping was already replaced due FD reuse, put back the stored Channel.
+            channels.put(fd, old);
+
+            // If we found another Channel in the map that is mapped to the same FD the given Channel MUST be closed.
+            assert !ch.isOpen();
+        }
+    }
+
+    private void closeAll() {
+        logger.trace("CloseAll IOUringEvenloop");
+        // Using the intermediate collection to prevent ConcurrentModificationException.
+        // In the `close()` method, the channel is deleted from `channels` map.
+        AbstractIOUringChannel[] localChannels = channels.values().toArray(new AbstractIOUringChannel[0]);
+
+        for (AbstractIOUringChannel ch : localChannels) {
+            ch.unsafe().close(ch.unsafe().voidPromise());
+        }
+    }
+
+    @Override
+    protected void run() {
+        final IOUringCompletionQueue completionQueue = ringBuffer.getIoUringCompletionQueue();
+        final IOUringSubmissionQueue submissionQueue = ringBuffer.getIoUringSubmissionQueue();
+
+        // Lets add the eventfd related events before starting to do any real work.
+        submissionQueue.addPollIn(eventfd.intValue());
+        submissionQueue.submit();
+
+        for (;;) {
+            logger.trace("Run IOUringEventLoop {}", this.toString());
+            long curDeadlineNanos = nextScheduledTaskDeadlineNanos();
+            if (curDeadlineNanos == -1L) {
+                curDeadlineNanos = NONE; // nothing on the calendar
+            }
+            nextWakeupNanos.set(curDeadlineNanos);
+
+            // Only submit a timeout if there are no tasks to process and do a blocking operation
+            // on the completionQueue.
+            if (!hasTasks()) {
+                try {
+                    if (curDeadlineNanos != prevDeadlineNanos) {
+                        prevDeadlineNanos = curDeadlineNanos;
+                        submissionQueue.addTimeout(deadlineToDelayNanos(curDeadlineNanos));
```

Comment:
From what I can see in the doc, the default behaviour of `IORING_OP_TIMEOUT` is to also complete if any other event completes first, i.e. it won't persist across calls to io_uring_enter with GET_EVENTS.

So something might need to be tweaked, maybe just to pass a very large value for the `len` field which should make timeouts stick until expired or cancelled.